File: 2025/05 May/30 Closest Neighbour in BST/main.py
# Link: https://www.geeksforgeeks.org/problems/closest-neighbor-in-bst/1


class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


def createBinarySearchTree(s: str):
    if not s or s == "N":
        return None
    values = s.split()
    root = Node(int(values[0]))
    queue = [root]
    i = 1
    while i < len(values):
        current = queue.pop(0)
        if values[i] != "N":
            current.left = Node(int(values[i]))
            queue.append(current.left)
        i += 1
        if i < len(values) and values[i] != "N":
            current.right = Node(int(values[i]))
            queue.append(current.right)
        i += 1
    return root


class Solution:
    def findMaxFork(self, root, k):
        if root == None:
            return -1
        if root.data == k:
            return k
        elif root.data < k:
            ans = self.findMaxFork(root.right, k)
            if ans == -1:
                return root.data
            else:
                return ans
        elif root.data > k:
            return self.findMaxFork(root.left, k)


if __name__ == "__main__":
    obj = Solution()
    s = "10 7 15 2 8 11 16"
    k = 14
    root = createBinarySearchTree(s)
    ans = obj.findMaxFork(root, k)
    print(ans)
    s = "5 2 12 1 3 9 21 N N N N N N 19 25"
    k = 24
    root = createBinarySearchTree(s)
    ans = obj.findMaxFork(root, k)
    print(ans)
    s = "5 2 12 1 3 9 21 N N N N N N 19 25"
    k = 4
    root = createBinarySearchTree(s)
    ans = obj.findMaxFork(root, k)
    print(ans)

File: 2025/05 May/30 Closest Neighbour in BST/main.cpp
// Link: https://www.geeksforgeeks.org/problems/closest-neighbor-in-bst/1

#include <iostream>
using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int x)
    {
        data = x;
        left = NULL;
        right = NULL;
    }
};

Node *createBinarySearchTree(string str)
{
    Node *root = NULL;
    int n = str.length();
    int i = 0;
    while (i < n)
    {
        if (str[i] == 'N')
        {
            i += 2;
            continue;
        }
        int num = 0;
        while (i < n && str[i] != ' ')
        {
            num = num * 10 + (str[i] - '0');
            i++;
        }
        Node *newNode = new Node(num);
        if (!root)
        {
            root = newNode;
        }
        else
        {
            Node *current = root;
            while (true)
            {
                if (num < current->data)
                {
                    if (!current->left)
                    {
                        current->left = newNode;
                        break;
                    }
                    current = current->left;
                }
                else
                {
                    if (!current->right)
                    {
                        current->right = newNode;
                        break;
                    }
                    current = current->right;
                }
            }
        }
        i++;
    }
    return root;
}

class Solution
{
public:
    int findMaxFork(Node *root, int k)
    {
        if (root == NULL)
            return -1;
        if (root->data == k)
            return k;

        else if (root->data < k)
        {

            int x = findMaxFork(root->right, k);
            if (x == -1)
                return root->data;
            else
                return x;
        }
        return findMaxFork(root->left, k);
    }
};

int main()
{
    Solution obj;
    string str = "10 7 15 2 8 11 16";
    int k = 14;
    Node *root = createBinarySearchTree(str);
    int result = obj.findMaxFork(root, k);
    if (result == -1)
        cout << "No closest neighbour found." << endl;
    else
        cout << "Closest neighbour to " << k << " is: " << result << endl;

    str = "5 2 12 1 3 9 21 N N N N N N 19 25";
    k = 24;
    root = createBinarySearchTree(str);
    result = obj.findMaxFork(root, k);
    if (result == -1)
        cout << "No closest neighbour found." << endl;
    else
        cout << "Closest neighbour to " << k << " is: " << result << endl;

    str = "5 2 12 1 3 9 21 N N N N N N 19 25";
    k = 4;
    root = createBinarySearchTree(str);
    result = obj.findMaxFork(root, k);
    if (result == -1)
        cout << "No closest neighbour found." << endl;
    else
        cout << "Closest neighbour to " << k << " is: " << result << endl;
    return 0;
}